governance/api.go

```diff
@@ -146,22 +162,166 @@ func (api *GovernanceKlayAPI) GetRewards(num *rpc.BlockNumber) (*reward.RewardSp
 	return reward.GetBlockReward(header, rules, rewardParamSet)
 }
 
-func (api *GovernanceKlayAPI) ChainConfig() *params.ChainConfig {
-	num := rpc.LatestBlockNumber
-	return getChainConfig(api.governance, &num)
-}
+type accumulatedRewards struct {
+	Period         uint     `json:"period"`
+	FirstBlockTime string   `json:"firstBlockTime"`
+	LastBlockTime  string   `json:"lastBlockTime"`
+	FirstBlock     *big.Int `json:"firstBlock"`
+	LastBlock      *big.Int `json:"lastBlock"`
+
+	// TotalMinted + TotalTxFee - TotalBurntTxFee = TotalProposerRewards + TotalStakingRewards + TotalKFFRewards + TotalKCFRewards
+	TotalMinted          *big.Int                    `json:"totalMinted"`
+	TotalTxFee           *big.Int                    `json:"totalTxFee"`
+	TotalBurntTxFee      *big.Int                    `json:"totalBurntTxFee"`
+	TotalProposerRewards *big.Int                    `json:"totalProposerRewards"`
+	TotalStakingRewards  *big.Int                    `json:"totalStakingRewards"`
+	TotalKFFRewards      *big.Int                    `json:"totalKFFRewards"`
+	TotalKCFRewards      *big.Int                    `json:"totalKCFRewards"`
+	Rewards              map[common.Address]*big.Int `json:"rewards"`
+}
+
+// GetRewardsAccumulated returns accumulated rewards data in the range of [start, end).
+// The given block range is divided into many accumulation periods with a given period parameter.
+func (api *GovernanceAPI) GetRewardsAccumulated(start rpc.BlockNumber, end rpc.BlockNumber, period uint64) ([]accumulatedRewards, error) {
+	type request struct {
+		period   uint64
+		blockNum uint64
+	}
 
-// TODO-Klaytn-Mantle: deprecate this
-func (api *GovernanceKlayAPI) ChainConfigAt(num *rpc.BlockNumber) *params.ChainConfig {
-	return getChainConfig(api.governance, num)
-}
+	blockchain := api.governance.BlockChain()
+	govKlayAPI := NewGovernanceKlayAPI(api.governance, blockchain)
 
-func (api *GovernanceKlayAPI) GetChainConfig(num *rpc.BlockNumber) *params.ChainConfig {
-	return getChainConfig(api.governance, num)
+	currentBlock := blockchain.CurrentBlock().NumberU64()
+
+	startBlock := currentBlock
+	if start >= rpc.EarliestBlockNumber {
+		startBlock = uint64(start.Int64())
+	}
+
+	endBlock := currentBlock
+	if end >= rpc.EarliestBlockNumber {
+		endBlock = uint64(end.Int64())
+	}
+
+	if period <= 0 {
+		period = 86400 // one day
+	}
+
+	if startBlock >= endBlock {
+		return nil, errors.New("the end block number should be larger the start block number")
+	}
+
+	if endBlock > currentBlock {
+		return nil, errors.New("the end block number should be equal or less than the current block number")
+	}
+
+	blockCount := endBlock - startBlock
+	if blockCount > 604800 { // 7 days. naive resource protection
```

Comment:
We might have a global lock or semaphore for this API

Reply:
I implemented this API in `governance` API that is not usually open for public. Do you still think the lock is required?

Reply:
It was a suggestion for alternative resource protection. I don't think we need a lock if it's in governance namespace.